### ISSUE TITLE
Deletion of upgrade-health-check during the upgrade preflight could fail  due to a race condition and surfaced as CreateJob failure

### ISSUE THREAD

**Issue by yuyabee** [kubernetes/kubernetes#122303]
### What happened?

I have seen false positives in the upgrade preflight where `CreateJob` fails because `kube-system/upgrade-health-check` job is not found when trying to delete it.

Looking at the code, there is a bit of inconsistent treatment in the creation and the deletion of the job during the cluster health check.

When creating the job to test if a job can be created on the cluster, [the job creation accounts for a possibility of race condition and have a retry to mitigate this](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/upgrade/health.go#L154-L163).

However, the deletion code doesn't take care of the same race condition in two places:
1. https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/upgrade/health.go#L142-L147
2. https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/upgrade/health.go#L149-L152

Specifically, I think [deleteHealthCheckJob](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/upgrade/health.go#L197-L204) should ignore `NotFound` errors to be resilient to the cases where the job was deleted by another client (basically `--ignore-not-found=true` is what we want here).

### What did you expect to happen?

Ignore `NotFound` error when attempting to delete the `upgrade-health-check` job.

### How can we reproduce it (as minimally and precisely as possible)?

With 3 control plane node setup and run `kubeadm upgrade plan` on three nodes at the same time and this probably happens once in a while.

### Anything else we need to know?

If the requested change sounds reasonable, I can work on making the changes.

### Kubernetes version

<details>

```console
$ kubectl version
Have observed this in 1.25.x to 1.27.x
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by pacoxu** [comment]
/cc @neolit123 
Do we support `run kubeadm upgrade plan on three nodes at the same time`?

**Comment by neolit123** [comment]
> /cc @neolit123 Do we support `run kubeadm upgrade plan on three nodes at the same time`?

we do not. please call "upgrade plan" only once - only on the node where you run "upgrade apply".
this is already in the docs.

**Comment by yuyabee** [comment]
Running upgrade plan on multiple nodes was just an easy repro process I thought of, but this issue exists if the `CheckClusterHealth` is called from different places and running at the same time.

Basically [the scenario in which the job can be getting deleted while trying to create](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/upgrade/health.go#L154) is the exact scenario that could break the preflight and return error introduced by [this commit](https://github.com/kubernetes/kubernetes/commit/906d3159b120ed56c0ebc89e50ecb0cfa9533079#diff-a8d816705b71f0f0af0f5736a64c04ba1494d678d95cff4b2781533ae780234bR156).

Even in this case, we still don't want to make it more reliable by simply ignoring the job `NotFound` when trying to delete it?

**Comment by neolit123** [comment]
> Even in this case, we still don't want to make it more reliable by simply ignoring the job NotFound when trying to delete it?

the delete code is being removed:
https://github.com/kubernetes/kubernetes/pull/122079

this will land in 1.30.
i don't think we should backport a fix. IMO, just avoid calling plan for multiple places, that will never be supported.

**Comment by taotriad** [comment]
This is, at the very least, trivial to trigger even with kubeadm 1.30.0 by doing:

`sudo kubeadm upgrade plan 1.30.0 && sudo kubeadm upgrade apply 1.30.0`

It's possible to workaround this by using `kubeadm upgrade apply --ignore-preflight-errors=CreateJob` (should be safe since the check is already performed during plan), but it seems to me that having a known race condition that is bound to trigger anyone who is scripting their upgrades (Ansible jobs or similar) isn't an ideal scenario.

**Comment by neolit123** [comment]
what errors are you getting @taotriad ?

cc @carlory

**Comment by taotriad** [comment]
Upgrading one of my clusters gave a CreateJob failure from HealthCheckJob when running the second command in my previous comment (`kubeadm upgrade apply 1.30.0`), fully repeatable (until I added `--ignore-preflight-errors=CreateJob`; now the cluster is already upgraded, so it's obviously impossible to reproduce it...)

The other clusters where I've tested this so far upgraded successfully. It seems to be a function of whether or not the upgrade-heath-check Job from `kubeadm upgrade plan` has disappeared before `kubeadm upgrade apply` starts. Why this happens on one control plane but not others I cannot answer. The clusters run different generation hardware, different CNI, workloads, etc., so tracking down the trigger might be difficult, but it should still be fairly simple to solve by having `kubeadm upgrade apply` wait for any previous health-check jobs to disappear before trying to create a new one (or simply run health-check from *either* plan or apply, not both).

**Comment by carlory** [comment]
Is the ttl too short? when polling the job,  but it is deleted? Would it be possible to increase its value to 5?

https://github.com/kubernetes/kubernetes/blob/16a594f907d0d4a6224dab2d0704793d5e7898f6/cmd/kubeadm/app/phases/upgrade/health.go#L117

cc @neolit123

**Comment by neolit123** [comment]
@taotriad what is the kubeadm version where you saw the problem? (version to upgrade to)
in 1.30 this should not happen:
https://github.com/kubernetes/kubernetes/commit/c62155a7705d8dc0302d1b4c55e645904583f8c9

- jobs have unique names
- they are auto-deleted

**Comment by geotransformer** [comment]
Still observed this issue in 1.31.2

    set -o pipefail
    kubeadm upgrade plan --config=/etc/kubernetes/kubeadmin-config.yaml
    echo "Y" | kubeadm upgrade apply --config=/etc/kubernetes/kubeadmin-config.yaml 1.31.2
 
    [upgrade/health] FATAL: [preflight] Some fatal errors occurred:
            [ERROR CreateJob]: Job "upgrade-health-check-4qjpt" in the namespace "kube-system" did not complete in 15s: no condition of type Complete
    [preflight] If you know what you are doing, you can make a check non-fatal with `--ignore-preflight-errors=...`
    To see the stack trace of this error execute with --v=5 or higher
    
    [upgrade/health] FATAL: [preflight] Some fatal errors occurred:
            [ERROR CreateJob]: Job "upgrade-health-check-kkvjt" in the namespace "kube-system" did not complete in 15s: no condition of type Complete
    [preflight] If you know what you are doing, you can make a check non-fatal with `--ignore-preflight-errors=...`
    To see the stack trace of this error execute with --v=5 or higher

**Comment by yuyabee** [comment]
@geotransformer  I don't think the error you reported is the same one as the one I originally reported. The original issue was in the code to delete the upgrade preflight job. However the code to delete the job itself was deleted in 1.30 in favor of using TTL (as pointed out in https://github.com/kubernetes/kubernetes/issues/122303#issuecomment-1856012987). So, in 1.31 this original issue is really not reproducible. 

What's failing in the log you provided is `CreateJob` step, so you need to check the job to see exactly why it failed.

**Comment by mahendraredwarasila-snc** [comment]
> Still observed this issue in 1.31.2
> 
> ```
> set -o pipefail
> kubeadm upgrade plan --config=/etc/kubernetes/kubeadmin-config.yaml
> echo "Y" | kubeadm upgrade apply --config=/etc/kubernetes/kubeadmin-config.yaml 1.31.2
> 
> [upgrade/health] FATAL: [preflight] Some fatal errors occurred:
>         [ERROR CreateJob]: Job "upgrade-health-check-4qjpt" in the namespace "kube-system" did not complete in 15s: no condition of type Complete
> [preflight] If you know what you are doing, you can make a check non-fatal with `--ignore-preflight-errors=...`
> To see the stack trace of this error execute with --v=5 or higher
> 
> [upgrade/health] FATAL: [preflight] Some fatal errors occurred:
>         [ERROR CreateJob]: Job "upgrade-health-check-kkvjt" in the namespace "kube-system" did not complete in 15s: no condition of type Complete
> [preflight] If you know what you are doing, you can make a check non-fatal with `--ignore-preflight-errors=...`
> To see the stack trace of this error execute with --v=5 or higher
> ```

We did come across same issue while upgrading to 1.31.4

**Comment by neolit123** [comment]
users hitting that issue could have cluster health issues.


> [@geotransformer](https://github.com/geotransformer), can you provide the yaml and logs file for that pod? It would be helpful to investigate the issue.

+1

**Comment by mahendraredwarasila-snc** [comment]
> users hitting that issue could have cluster health issues.
> 
> > can you provide the yaml and logs file for that pod? It would be helpful to investigate the issue.
> 
> +1

Here is some logs from the kube-controller-manager which has errors 

```
# kubectl -n kube-system logs kube-controller-manager-k8s-mtest016-control01.example.com | grep upgrade-health-check-xtn5f
I0219 12:55:08.739155       1 job_controller.go:568] "enqueueing job" logger="job-controller" key="kube-system/upgrade-health-check-xtn5f" delay="0s"
I0219 12:55:08.746045       1 job_controller.go:568] "enqueueing job" logger="job-controller" key="kube-system/upgrade-health-check-xtn5f" delay="1s"
I0219 12:55:08.751169       1 job_controller.go:568] "enqueueing job" logger="job-controller" key="kube-system/upgrade-health-check-xtn5f" delay="1s"
I0219 12:55:08.751567       1 job_controller.go:568] "enqueueing job" logger="job-controller" key="kube-system/upgrade-health-check-xtn5f" delay="1s"
I0219 12:55:08.763235       1 job_controller.go:568] "enqueueing job" logger="job-controller" key="kube-system/upgrade-health-check-xtn5f" delay="1s"
I0219 12:55:10.083312       1 job_controller.go:568] "enqueueing job" logger="job-controller" key="kube-system/upgrade-health-check-xtn5f" delay="1s"
I0219 12:55:11.210767       1 job_controller.go:568] "enqueueing job" logger="job-controller" key="kube-system/upgrade-health-check-xtn5f" delay="1s"
E0219 12:55:12.216007       1 job_controller.go:601] "Unhandled Error" err="syncing job: tracking status: adding uncounted pods to status: Job.batch \"upgrade-health-check-xtn5f\" is invalid: [status.conditions: Invalid value: cannot set SuccessCriteriaMet to NonIndexed Job, status.conditions: Invalid value: cannot set SuccessCriteriaMet=True for Job without SuccessPolicy]" logger="UnhandledError"
E0219 12:55:13.222589       1 job_controller.go:601] "Unhandled Error" err="syncing job: tracking status: adding uncounted pods to status: Job.batch \"upgrade-health-check-xtn5f\" is invalid: [status.conditions: Invalid value: cannot set SuccessCriteriaMet to NonIndexed Job, status.conditions: Invalid value: cannot set SuccessCriteriaMet=True for Job without SuccessPolicy]" logger="UnhandledError"
E0219 12:55:15.229341       1 job_controller.go:601] "Unhandled Error" err="syncing job: tracking status: adding uncounted pods to status: Job.batch \"upgrade-health-check-xtn5f\" is invalid: [status.conditions: Invalid value: cannot set SuccessCriteriaMet to NonIndexed Job, status.conditions: Invalid value: cannot set SuccessCriteriaMet=True for Job without SuccessPolicy]" logger="UnhandledError"
E0219 12:55:19.236106       1 job_controller.go:601] "Unhandled Error" err="syncing job: tracking status: adding uncounted pods to status: Job.batch \"upgrade-health-check-xtn5f\" is invalid: [status.conditions: Invalid value: cannot set SuccessCriteriaMet to NonIndexed Job, status.conditions: Invalid value: cannot set SuccessCriteriaMet=True for Job without SuccessPolicy]" logger="UnhandledError"
E0219 12:55:27.243241       1 job_controller.go:601] "Unhandled Error" err="syncing job: tracking status: adding uncounted pods to status: Job.batch \"upgrade-health-check-xtn5f\" is invalid: [status.conditions: Invalid value: cannot set SuccessCriteriaMet to NonIndexed Job, status.conditions: Invalid value: cannot set SuccessCriteriaMet=True for Job without SuccessPolicy]" logger="UnhandledError"
I0219 12:55:28.006063       1 job_controller.go:568] "enqueueing job" logger="job-controller" key="kube-system/upgrade-health-check-xtn5f" delay="1s"
I0219 12:55:28.012476       1 job_controller.go:568] "enqueueing job" logger="job-controller" key="kube-system/upgrade-health-check-xtn5f" delay="1s"
I0219 12:55:28.017081       1 job_controller.go:568] "enqueueing job" logger="job-controller" key="kube-system/upgrade-health-check-xtn5f" delay="1s"
I0219 12:55:29.581566       1 job_controller.go:568] "enqueueing job" logger="job-controller" key="kube-system/upgrade-health-check-xtn5f" delay="1s"

```

this failure only occurs then the kube-controller-manager lease is on the controller that has been upgrade to 1.31.x (i.e. does not happen if the lease is on nodes that have not been upgraded to kube-controller-manager 1.30.x).

Please also let me know if we need to create a new issue for this. Thanks

**Comment by neolit123** [comment]
> Invalid value: cannot set SuccessCriteriaMet=True for Job without SuccessPolicy

any idea what that means?
is there anything we can add to the Job object when kubeadm creates it?

> this failure only occurs then the kube-controller-manager lease is on the controller that has been upgrade to 1.31.x (i.e. does not happen if the lease is on nodes that have not been upgraded to kube-controller-manager 1.30.x).

sounds like a kcm bug.

**Comment by pacoxu** [comment]
> > Invalid value: cannot set SuccessCriteriaMet=True for Job without SuccessPolicy
> 
> any idea what that means? is there anything we can add to the Job object when kubeadm creates it?

It seems to be related to https://github.com/kubernetes/enhancements/issues/3998.  @tenzen-y @alculquicondor do you have some comments on this?

**Comment by pacoxu** [comment]
> > this failure only occurs then the kube-controller-manager lease is on the controller that has been upgrade to 1.31.x (i.e. does not happen if the lease is on nodes that have not been upgraded to kube-controller-manager 1.30.x).
> 
> sounds like a kcm bug.

/reopen

**Comment by carlory** [comment]
Do you disable the feature-gate `JobPodReplacementPolicy` manually? @mahendraredwarasila-snc 

It seems related to https://github.com/kubernetes/kubernetes/commit/fb7704ba038ce825c64fe5313cb6addbc7a7e3d6#diff-a7577a7dede7864ff38c631319033714fdd0bed91108d976282e1099507e6ff0R988. 

It always sets the condition type `JobSuccessCriteriaMet` even if the job is NonIndexed. cc @mimowo

**Comment by mahendraredwarasila-snc** [comment]
> JobPodReplacementPolicy

I have tried this changing it manually but i couldn't find success with it

```
kubectl get cm kubeadm-config  -n kube-system -o yaml | grep -C 2 featureGates 
        - etcd.kube-system
        - etcd
    featureGates:
      ControlPlaneKubeletLocalMode: true
      JobPodReplacementPolicy: false
```

**Comment by carlory** [comment]
> this failure only occurs then the kube-controller-manager lease is on the controller that has been upgrade to 1.31.x (i.e. does not happen if the lease is on nodes that have not been upgraded to kube-controller-manager 1.30.x).

I think we need to backport https://github.com/kubernetes/kubernetes/pull/125510 to release-1.30 because the newer controller can not work well with the older kube-apiserver. cc @mimowo

**Comment by mimowo** [comment]
I haven't caught up yet with the thread fully, but the errors `cannot set SuccessCriteriaMet to NonIndexed Job` indicate that the kube-controller-manager in 1.31 tries to set the SuccessCriteriaMet condtion whilst the validation code  (api-server) does not allow for it, meaning some inconsistent state, potential options I see:
1. a bug in k8s that the validation does not match controller logic
2. the kube-apiserver is temporarily older than kube-controller-manager. Such a skew policy is not supported: https://kubernetes.io/releases/version-skew-policy/#kube-controller-manager-kube-scheduler-and-cloud-controller-manager.
3. the enablement of the feature gates is inconsistent between the kube-apiserver and kube-controller-manager

I would suggest first eliminate 2 and 3: make sure the kube-apiserver version is >= kube-controller-manager, and make sure `JobPodReplacementPolicy` is consistently enabled in both kube-apisever and kube-controller manager. 

Finally, if you have control over the created Job, then using an indexed job should workaround the validation. Is this an option?

**Comment by mimowo** [comment]
> I think we need to backport https://github.com/kubernetes/kubernetes/pull/125510 to release-1.30 because the newer controller can not work well with the older kube-apiserver. cc @mimowo

I don't think this is a possible option, because we don't backport features. 

Looking at the validation code: https://github.com/kubernetes/kubernetes/blob/566f939b19fa0be3d3f5d11b13c38884887eb915/pkg/apis/batch/validation/validation.go#L590-L592

It is disabled by `AllowForSuccessCriteriaMetInExtendedScope` which on 1.31 can be turned on by enabling JobReplacementPolicy (should be on by default): https://github.com/kubernetes/kubernetes/blob/11fefab875b49caca7a10c23dc31b60c16ab8420/pkg/registry/batch/job/strategy.go#L411-L415

So I'm wondering what is the version of kube-apiserver when you observe the failures (1.30 vs 1.31)? Try to make sure JobReplacementPolicy is enabled on the kube-apiserver.

**Comment by neolit123** [comment]
> > JobPodReplacementPolicy
> 
> I have tried this changing it manually but i couldn't find success with it
> 
> ```
> kubectl get cm kubeadm-config  -n kube-system -o yaml | grep -C 2 featureGates 
>         - etcd.kube-system
>         - etcd
>     featureGates:
>       ControlPlaneKubeletLocalMode: true
>       JobPodReplacementPolicy: false
> ```

the feature gate should not be under the kubeadm featureGates.
https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-init/#feature-gates
it should be under k8s component extraArgs / --feature-gates flags.
https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/control-plane-flags

**Comment by neolit123** [comment]
> the kube-apiserver is temporarily older than kube-controller-manager. Such a skew policy is not supported: https://kubernetes.io/releases/version-skew-policy/#kube-controller-manager-kube-scheduler-and-cloud-controller-manager.

on an upgraded cp node, kubeadm writes updated kcm and apiserver static pods manifests and then the kubelet restarts them at the same time on the next manifest folder read. supposedly there should not be this skew.

> the enablement of the feature gates is inconsistent between the kube-apiserver and kube-controller-manager

@mahendraredwarasila-snc please confirm.

**Comment by tenzen-y** [comment]
> > > Invalid value: cannot set SuccessCriteriaMet=True for Job without SuccessPolicy
> > 
> > 
> > any idea what that means? is there anything we can add to the Job object when kubeadm creates it?
> 
> It seems to be related to [kubernetes/enhancements#3998](https://github.com/kubernetes/enhancements/issues/3998). [@tenzen-y](https://github.com/tenzen-y) [@alculquicondor](https://github.com/alculquicondor) do you have some comments on this?

ACK

**Comment by tenzen-y** [comment]
> > I think we need to backport [#125510](https://github.com/kubernetes/kubernetes/pull/125510) to release-1.30 because the newer controller can not work well with the older kube-apiserver. cc [@mimowo](https://github.com/mimowo)
> 
> I don't think this is a possible option, because we don't backport features.
> 
> Looking at the validation code:
> 
> [kubernetes/pkg/apis/batch/validation/validation.go](https://github.com/kubernetes/kubernetes/blob/566f939b19fa0be3d3f5d11b13c38884887eb915/pkg/apis/batch/validation/validation.go#L590-L592)
> 
> Lines 590 to 592 in [566f939](/kubernetes/kubernetes/commit/566f939b19fa0be3d3f5d11b13c38884887eb915)
> 
>  if !opts.AllowForSuccessCriteriaMetInExtendedScope && ptr.Deref(job.Spec.CompletionMode, batch.NonIndexedCompletion) != batch.IndexedCompletion && isJobSuccessCriteriaMet(job) { 
>  	allErrs = append(allErrs, field.Invalid(fldPath.Child("conditions"), field.OmitValueType{}, "cannot set SuccessCriteriaMet to NonIndexed Job")) 
>  } 
> It is disabled by `AllowForSuccessCriteriaMetInExtendedScope` which on 1.31 can be turned on by enabling JobReplacementPolicy (should be on by default):
> 
> [kubernetes/pkg/registry/batch/job/strategy.go](https://github.com/kubernetes/kubernetes/blob/11fefab875b49caca7a10c23dc31b60c16ab8420/pkg/registry/batch/job/strategy.go#L411-L415)
> 
> Lines 411 to 415 in [11fefab](/kubernetes/kubernetes/commit/11fefab875b49caca7a10c23dc31b60c16ab8420)
> 
>  if utilfeature.DefaultFeatureGate.Enabled(features.JobPodReplacementPolicy) { 
>  	return batchvalidation.JobStatusValidationOptions{ 
>  		AllowForSuccessCriteriaMetInExtendedScope: true, 
>  	} 
>  } 
> So I'm wondering what is the version of kube-apiserver when you observe the failures (1.30 vs 1.31)? Try to make sure JobReplacementPolicy is enabled on the kube-apiserver.

Thank you for summarizing that. I'm with @mimowo here.
As @mimowo mentioned, I guess the root cause is the disabled `JobPodReplacementPolicy` (enabled by default in v1.29, see:https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#overview) during the situations where there are version gaps (cluster upgrading) between kube-api-server (1.30) and kube-controller-manager (1.31). So, I suspect that you tried to directly jump cluster version from 1.28 to 1.30 (or 1.31) or manually disable `JobPodReplacementPolicy`.

**Comment by neolit123** [comment]
> > and then the kubelet restarts them at the same time on the next manifest folder read
> 
> Depends what "at the same time" means in this context. Is it possible that temporarily one is running old while another new? If this is the case then the skew policy requires the kube-apiserver to be upgraded (restarted in this case, IIUC) first.

actually this could end up being problematic if this fg is sensitive to the skew and doesn't not handle well the error i.e. controller doesn't resolve the bad state after a while when the skew is not there anymore.

it iterates over the list of component and calls updateComponent
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/upgrade/staticpods.go#L496
but the function can block for n seconds.

in the meantime the kubelet static pod reader is cycling for about 40 seconds iirc.
so if apiserver gets updated in one cycle and the other components miss this cycle there could be a temporary skew.

we could make the upgrade of components concurrent, which i don't like as the logs end up noisy and it needs good handling of errors, or this constant could be changed to have the apiserver last
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/constants/constants.go#L485

@pacoxu @carlory

**Comment by tenzen-y** [comment]
> > > > I think we need to backport [#125510](https://github.com/kubernetes/kubernetes/pull/125510) to release-1.30 because the newer controller can not work well with the older kube-apiserver. cc [@mimowo](https://github.com/mimowo)
> > > 
> > > 
> > > I don't think this is a possible option, because we don't backport features.
> > > Looking at the validation code:
> > > [kubernetes/pkg/apis/batch/validation/validation.go](https://github.com/kubernetes/kubernetes/blob/566f939b19fa0be3d3f5d11b13c38884887eb915/pkg/apis/batch/validation/validation.go#L590-L592)
> > > Lines 590 to 592 in [566f939](/kubernetes/kubernetes/commit/566f939b19fa0be3d3f5d11b13c38884887eb915)
> > > if !opts.AllowForSuccessCriteriaMetInExtendedScope && ptr.Deref(job.Spec.CompletionMode, batch.NonIndexedCompletion) != batch.IndexedCompletion && isJobSuccessCriteriaMet(job) {
> > > allErrs = append(allErrs, field.Invalid(fldPath.Child("conditions"), field.OmitValueType{}, "cannot set SuccessCriteriaMet to NonIndexed Job"))
> > > }
> > > It is disabled by `AllowForSuccessCriteriaMetInExtendedScope` which on 1.31 can be turned on by enabling JobReplacementPolicy (should be on by default):
> > > [kubernetes/pkg/registry/batch/job/strategy.go](https://github.com/kubernetes/kubernetes/blob/11fefab875b49caca7a10c23dc31b60c16ab8420/pkg/registry/batch/job/strategy.go#L411-L415)
> > > Lines 411 to 415 in [11fefab](/kubernetes/kubernetes/commit/11fefab875b49caca7a10c23dc31b60c16ab8420)
> > > if utilfeature.DefaultFeatureGate.Enabled(features.JobPodReplacementPolicy) {
> > > return batchvalidation.JobStatusValidationOptions{
> > > AllowForSuccessCriteriaMetInExtendedScope: true,
> > > }
> > > }
> > > So I'm wondering what is the version of kube-apiserver when you observe the failures (1.30 vs 1.31)? Try to make sure JobReplacementPolicy is enabled on the kube-apiserver.
> > 
> > 
> > Thank you for summarizing that. I'm with [@mimowo](https://github.com/mimowo) here. As [@mimowo](https://github.com/mimowo) mentioned, I guess the root cause is the disabled `JobPodReplacementPolicy` (enabled by default in v1.29, see:https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#overview) during the situations where there are version gaps (cluster upgrading) between kube-api-server (1.30) and kube-controller-manager (1.31). So, I suspect that you tried to directly jump cluster version from 1.28 to 1.30 (or 1.31) or manually disable `JobPodReplacementPolicy`.
> 
> We have not disabled the JobPodReplacementPolicy and are attempting to upgrade from version 1.30.8 to 1.31.4. We also have set ControlPlaneKubeletLocalMode: true because our cluster is behind a VIP and we do not want the kubelet to communicate with an older version of the API server. I suspect this might be causing the issue, as the kubelet is pointing to the local API server, and the KCM is a newer version where the node it is running on has already been upgraded. I'm not sure how would we handle this?

Thank you for describing your situation. IIRC, there are some `ControlPlaneKubeletLocalMode` related problems like this: https://github.com/kubernetes/kubernetes/issues/127316

Actually, I have the experience to hit #12316 problems in my cluster. At that time, I manually stopped the old version kube-apiserver to resolve the conflicts between kube-controller-manager version. I'm not sure if this could resolve your situation and you face the same situation.

**Comment by mahendraredwarasila-snc** [comment]
> ```
>              master 1      master 2          master 3
> 
> 
>           +-------+      +--------+       +--------+
> APISERVER |  1.31 |      |  1.31  |    +->+  1.30  |
>           +-------+      +--------+    |  +--------+
>                                        |
> VIP       +------------^---------------+------------+
>                        |
>           +-------+----+  +-------+       +--------+
> KCM       | 1.31  |       |       |       |   1.30 |
>           |       |       |   1.31|       |        |
>           +-------+       +-------+       +--------+
> ```
> 
> 1. the controller on master 1 become leader during master 2 is upgraded.
> 2. we will get a preflight error when upgrading master3

Yes, In fact we are getting health check  error for second node itself since the controller on master 1 become leader during master 2 is upgraded

**Comment by mimowo** [comment]
IIUC This diagram shows master3 apiserver in 1.30 and master 1 KCM in 1.31 which is the non-supported skew. 

So, I don't see this indicative of a bug in the core k8s (at least yet).

Maybe the easiest workaround is indeed to ensure JobReplacementPolicy=False on KCM instances during the upgrade. Then, the JobSuccessCriteriaMet condition will not be attempted to be added. Maybe comment https://github.com/kubernetes/kubernetes/issues/122303#issuecomment-2681129486 helps?

**Comment by carlory** [comment]
> we could make the upgrade of components concurrent, which i don't like as the logs end up noisy and it needs good handling of errors, or this constant could be changed to have the apiserver last

@neolit123 It won't fix this issue in HA mode. In HA mode, the controller is talking with apiserver via a vip. Kubeadm can't control the network path between the controller and apiserver.

**Comment by neolit123** [comment]
> > we could make the upgrade of components concurrent, which i don't like as the logs end up noisy and it needs good handling of errors, or this constant could be changed to have the apiserver last
> 
> [@neolit123](https://github.com/neolit123) It won't fix this issue in HA mode. In HA mode, the controller is talking with apiserver via a vip. Kubeadm can't control the network path between the controller and apiserver.



note that kubeadm's kcm's on cp nodes talk only to the local apiserver. kcm wouldn't go over the vip to talk to another apiserver.

**Comment by neolit123** [comment]
> Maybe the easiest workaround is indeed to ensure JobReplacementPolicy=False on KCM instances during the upgrade. Then, the JobSuccessCriteriaMet condition will not be attempted to be added. Maybe comment [#122303 (comment)](https://github.com/kubernetes/kubernetes/issues/122303#issuecomment-2681129486) helps?
> 
> ```
> the feature gate should not be under the kubeadm featureGates.
> https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-init/#feature-gates
> it should be under k8s component extraArgs / --feature-gates flags.
> https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/control-plane-flags
> ```

please try this @mahendraredwarasila-snc 


> or this constant could be changed to have the apiserver last https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/constants/constants.go#L485
> 
> [@pacoxu](https://github.com/pacoxu) [@carlory](https://github.com/carlory)

this is still a valid kubeadm fix.
but note that that component list is used in many places so best to run some e2e from our test repo on the pr that changes it.

**Comment by carlory** [comment]
@neolit123  kube-controller-manager, kube-scheduler, and cloud-controller-manager must not be newer than the kube-apiserver instances they communicate with. We can not move the apiserver to the last of the list.

**Comment by neolit123** [comment]
@carlory ok, right so kube-apiserver must be upgraded first and that's what's happening. i got confused.

@mimowo kubeadm is doing the upgrade correctly and point 2 of your post is not possible here:
https://github.com/kubernetes/kubernetes/issues/122303#issuecomment-2680844930

i don't think we have excluded point 1, while point 3 must be confirmed by @mahendraredwarasila-snc

**Comment by carlory** [comment]
@mahendraredwarasila-snc Could you check the kubeconfig of the controller manager pod? which ip address is it using to connect to the api server? I can not imagine why it is talking to the old apiserver.

**Comment by mahendraredwarasila-snc** [comment]
We use kubespray for cluster upgrade. I see, it's pointing to VIP for server field from conf. Also, i see kubespray uses  [kubeadm upgrade apply for first and other control plane nodes](https://github.com/kubernetes-sigs/kubespray/blob/master/roles/kubernetes/control-plane/tasks/kubeadm-upgrade.yml#L33) whereas kubernetes documentation refers to use  kubeadm upgrade apply  for first control plane node and kubeadm upgrade node for other control plane nodes. 

`apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: xxyxyx
    server: https://xyz.testexample.com:6443
  name: testcluster
contexts:
- context:
    cluster: e2eupgrade805741
    user: system:kube-controller-manager
  name: system:kube-controller-manager@testcluster
current-context: system:kube-controller-manager@testcluster
kind: Config
preferences: {}
users:
- name: system:kube-controller-manager
  user:
    client-certificate-data: xxyxyx
    client-key-data: xxx
`

**Comment by neolit123** [comment]
> We use kubespray for cluster upgrade. I see, it's pointing to VIP for server field from conf. Also, i see kubespray uses [kubeadm upgrade apply for first and other control plane nodes](https://github.com/kubernetes-sigs/kubespray/blob/master/roles/kubernetes/control-plane/tasks/kubeadm-upgrade.yml#L33) whereas kubernetes documentation refers to use kubeadm upgrade apply for first control plane node and kubeadm upgrade node for other control plane nodes. also it is intermittent and happens only when upgrading the second or third control plane of the cluster
> not for the first one.

that changes the story...
kubeadm defaults != kubespray defaults
so this is a kubespray issue, in the future please post in their issue tracker.

also this is a known kubespray problem
https://github.com/kubernetes-sigs/kubespray/issues/7821#issuecomment-2619029520
cc @VannTen 

/close
